src/Chapter2.hs

```diff
@@ -857,8 +875,33 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
 
+{-
+I have this solution, but it's not optimal.
+I tried to implement another, but it doesn't seem to work (see commented code below)
+-}
+rewind :: [a] -> [a]
+rewind [] = []
+rewind (x:xs) = rewind xs ++ [x]
+
+{-
+This function works fine for lists like [1..5]
+But it returns a reversed list of element indexes
+
+rewind :: [Int] -> [Int]
+rewind l = go (length l) l
+  where
+  go :: Int -> [Int] -> [Int]
+  go i l
+      | i >= 1 = i : go (i-1) (drop (i-1) l)
+      | otherwise = l
+
+I can't get how to create a list of elements themselves. I tried this code
+`i >= 1 = l!!i : go (i-1) (drop (i-1) l)`
+instead of
+`i >= 1 = i : go (i-1) (drop (i-1) l)`
+But I keep getting an error about the index being too large :(
```

Comment:
Yes, you are correct that the solution is not efficient. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator (you are right again!) and the recursive function that will do the addition at the start of the list which is instant!

Taking element by index in lists in Haskell is not a cheap operation, due to the way lists are implemented. You need to go through each element, until you meet the required position. That is why the solution with `!!` would not be efficient as well.

In Haskell, all such problems in lists are solved with the `:` cons and pattern matching.
Instead of the index, lets have as an accumulator the new list that we will construct on each recursive step.

```haskell
rewind :: [Int] -> [Int]
rewind l = go [] l
  where
    go :: [Int] -> [Int] -> [Int]
    go acc l
```

and now the most interesting begans.
Let's start our recursive function with the stop case. As we will go element by element, our final stop will be the empty list. In that case, we assume that the accumulator already has all elements reversed, so we return it:

```haskell
go acc [] = acc
```

Let's now look at the general case. Not would be the mind-blowing moment (at least it was for me 🤯 ) and we will look closer after:

```haskell
go acc (x:xs) = go (x:acc) xs
```

So what we are doing? We are adding the elements to the start of the accumulator.

Lets prove that it will be reversed in accumulator (even though we add to the beginning) on some example:

| acc | list |
|-----|----|
| [] | [1,2,3,4]|
| 1: [] | [2,3,4] |
| 2 : 1 : [] | [3,4] |
| 3 : 2 : 1 : [] | [4] |
| 4 : 3 : 2 : 1 : [] | [] |

So at the last step, the list is reversed, and the function works only by one walk-through the list!

Hope it makes sense, let me know if you have questions 🙂

Reply:
Thank you so much for the explanation! Yeah, it is a bit mind-blowing 😅 But understandable at the end. I have changed the function implementation to this:
```
rewind :: [Int] -> [Int]
rewind l = go [] l
  where
      go :: [Int] -> [Int] -> [Int]
      go acc [] = acc
      go acc (x:xs) = go (x:acc) xs
```
And it works perfectly!